reftable/block.c

```diff
@@ -0,0 +1,382 @@
+  int n = 0;
+
+  record_key(rec, &key);
+  n = encode_key(&restart, out, last, key, record_val_type(rec));
+  if (n < 0) {
+    goto err;
+  }
+  out.buf += n;
+  out.len -= n;
+
+  n = record_encode(rec, out, w->hash_size);
+  if (n < 0) {
+    goto err;
+  }
+
+  out.buf += n;
+  out.len -= n;
+
+  if (block_writer_register_restart(w, start.len - out.len, restart, key) < 0) {
+    goto err;
+  }
+
+  free(slice_yield(&key));
+  return 0;
+
+err:
+  free(slice_yield(&key));
+  return -1;
+}
+
+int block_writer_register_restart(struct block_writer *w, int n, bool restart,
+                                  struct slice key) {
+  int rlen = w->restart_len;
+  if (rlen >= MAX_RESTARTS) {
+    restart = false;
+  }
+
+  if (restart) {
+    rlen++;
+  }
+  if (2 + 3 * rlen + n > w->block_size - w->next) {
+    return -1;
+  }
+  if (restart) {
+    if (w->restart_len == w->restart_cap) {
+      w->restart_cap = w->restart_cap * 2 + 1;
+      w->restarts = realloc(w->restarts, sizeof(uint32_t) * w->restart_cap);
+    }
+
+    w->restarts[w->restart_len++] = w->next;
+  }
+
+  w->next += n;
+  slice_copy(&w->last_key, key);
+  w->entries++;
+  return 0;
+}
+
+int block_writer_finish(struct block_writer *w) {
+  for (int i = 0; i < w->restart_len; i++) {
+    put_u24(w->buf + w->next, w->restarts[i]);
+    w->next += 3;
+  }
+
+  put_u16(w->buf + w->next, w->restart_len);
+  w->next += 2;
+  put_u24(w->buf + 1 + w->header_off, w->next);
+
+  if (block_writer_type(w) == BLOCK_TYPE_LOG) {
+    int block_header_skip = 4 + w->header_off;
+    struct slice compressed = {};
+    uLongf dest_len = 0, src_len = 0;
+    slice_resize(&compressed, w->next - block_header_skip);
+
+    dest_len = compressed.len;
+    src_len = w->next - block_header_skip;
+
+    if (Z_OK != compress2(compressed.buf, &dest_len, w->buf + block_header_skip,
+                          src_len, 9)) {
+      free(slice_yield(&compressed));
+      return ZLIB_ERROR;
+    }
+    memcpy(w->buf + block_header_skip, compressed.buf, dest_len);
+    w->next = dest_len + block_header_skip;
+  }
+  return w->next;
+}
+
+byte block_reader_type(struct block_reader *r) {
+  return r->block.data[r->header_off];
+}
+
+int block_reader_init(struct block_reader *br, struct block *block,
+                      uint32_t header_off, uint32_t table_block_size,
+                      int hash_size) {
+  uint32_t full_block_size = table_block_size;
+  byte typ = block->data[header_off];
+  uint32_t sz = get_u24(block->data + header_off + 1);
+
+  if (!is_block_type(typ)) {
+    return FORMAT_ERROR;
+  }
+
+  if (typ == BLOCK_TYPE_LOG) {
+    struct slice uncompressed = {};
+    int block_header_skip = 4 + header_off;
+    uLongf dst_len = sz - block_header_skip;
+    uLongf src_len = block->len - block_header_skip;
+
+    slice_resize(&uncompressed, sz);
+    memcpy(uncompressed.buf, block->data, block_header_skip);
+
+    if (Z_OK != uncompress2(uncompressed.buf + block_header_skip, &dst_len,
```

Comment:
It seems that `uncompress2()` is not universally liked in the PR build. The `linux-clang` job claims that this is not declared.

It does appear to be a function that is declared correctly in the `zlib.h` header. If you include `git-compat-util.h` as I suggested earlier, this problem will go away.

Reply:
curious, as git-compat-util doesn't declare compress2.

When I compile against the compat header, I get:

```
reftable/reader.c: In function ‘parse_footer’:
reftable/reader.c:127:29: warning: implicit declaration of function ‘crc32’ [-Wimplicit-function-declaration]
     uint32_t computed_crc = crc32(0, footer, f - footer);
                             ^~~~~
```

Reply:
On CI, this also fails. Do you know how the rest of the source code consumes zlib functions?

Reply:
I just `git grep -w crc32`ed around to find out, and then `git grep 'zlib\.h'`ed some more, and it looks as if `cache.h` includes that header, not `git-compat-util.h`...

So you have two options now: either include `cache.h` instead of `git-compat-util.h`, or include `zlib.h` explicitly in your source files.

Reply:
so crc32 works now, but uncompress2 doesn't.

which version of zlib do you have (is there a minimum version?)

The linux-clang build says: 

  /usr/include/zlib.h:1197:21: note: 'uncompress' declared here

but I have a comment at 1197 in zlib.h

Reply:
This was a recent addition:

commit 37281ac222ee7ceb8cc3253f13c8fa26a88dd566
Author: Mark Adler <madler@alumni.caltech.edu>
Date:   Tue Nov 15 20:45:01 2016 -0700

    Add uncompress2() function, which returns the input size used.

I've included a bit of zlib source now, so this should now be fixed.